[1.0] VRM0.xのマテリアルの表情アニメーションに対応する

ke456-png opened the pull request:
Description

1.0でVRM0.xのモデルを読み込んだ際も表情アニメーションでマテリアルを変化できるように実装する.



- MaterialColor に対応
  -  [VRMC_vrm-1.0-beta/expressions.ja.md#materialcolorbind](https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_vrm-1.0-beta/expressions.ja.md#materialcolorbind)
      - color
      - emissionColor
      - shadeColor
      - rimColor
      - outlineColor

- TextureTransform に対応
  -  [VRMC_vrm-1.0-beta/expressions.ja.md#texturetransformbind](https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_vrm-1.0-beta/expressions.ja.md#texturetransformbind)
      - scale
      - offset

TextureTransform.RotationはVRMの仕様として表情に紐づけることが出来ない。

ke456-png committed:
renameMaterialPropertyをthree-vrm-coreへ移動

vrm0.xのマテリアルアニメーションに対応するためにrenameMaterialPropertyを移動する。

使用すると考えられる場所はvrm0.xでのマテリアルアニメーションをバインドする箇所だけだと考えられる

ke456-png committed:
VRM0.xのマテリアルアニメーションに対応する

読み込み時にVRM0.xの場合でもマテリアルを表情で変化させることが出来るようにする。

MaterialColor
- color
- emissionColor
- shadeColor
- rimColor
- outlineColor

TextureTransform
- scale
- offset

RotationはVRMの仕様として表情に紐づけることが出来ない。

0b5vr commented on packages/three-vrm-core/src/expressions/VRMExpressionLoaderPlugin.ts:
```diff
@@ -323,7 +325,57 @@ export class VRMExpressionLoaderPlugin implements GLTFLoaderPlugin {
+                if (Array.isArray(material)) {
+                  materials.push(
+                    ...material.filter(
+                      (mtl) => mtl.name === materialValue.materialName! && materials.indexOf(mtl) === -1,
```
ここのtype assertionいらなさそう
```suggestion
                      (mtl) => mtl.name === materialValue.materialName && materials.indexOf(mtl) === -1,
```

0b5vr replied:
あー違うわ、filter関数内のラムダだからビックリ必要なんだこれ

0b5vr commented on packages/three-vrm-core/src/expressions/VRMExpressionLoaderPlugin.ts:
```diff
@@ -323,7 +325,57 @@ export class VRMExpressionLoaderPlugin implements GLTFLoaderPlugin {
+            });
+
+            materials.forEach((material) => {
+              const expressionType = renameMaterialProperty(
+                materialValue.propertyName!,
+              ) as VRMExpressionMaterialColorType;
```
これで正しいexpressionTypeが取れる確証はありますか？

0b5vr commented on packages/three-vrm-core/src/expressions/VRMExpressionLoaderPlugin.ts:
```diff
@@ -323,7 +325,57 @@ export class VRMExpressionLoaderPlugin implements GLTFLoaderPlugin {
+                    targetValue: new THREE.Color(...materialValue.targetValue!.slice(0, 3)),
+                  }),
+                );
+              } else {
```
ここはelseで弾くよりかは、 `_MainTex_ST` と等価比較をしたくありませんか？

0b5vr replied:
および、ここがハズレた場合は、「このパラメータは対応してないよ」Warningを出すのが親切かなと思っています

0b5vr replied:
VRM0でMToonにアサインできるのはこれらですね。

![image](https://user-images.githubusercontent.com/7824814/184613112-12e297f7-7db6-4c62-9730-41667cb3bdb7.png)

0b5vr replied:
color, emissionColor, shadeColor, rimColor, outlineColor
すべてお名前は揃っていそうですね。

ただ、以下のようなmapを用意してやったほうがより良さそうと思います (i.e. `renameMaterialPropert` を用いない):

```
{
  '_Color': 'color',
  '_ShadeColor': 'shadeColor',
  ...
}
```

ke456-png committed:
renameMaterialPropertを使用せずVRM0用のMapを用いる

VRM0のMaterialColorTypeの名前からVRM1用に変換するMapを追加する

'_Color': 'color'

ke456-png committed:
MatrialBindの問題の修正

- OutlineColorのアニメーション対応
- _propertyNameMapMapの名前修正
- shadeTextureのUVアニメーションの修正

ke456-png committed:
表情アニメーションにバインドするマテリアルを配列にする箇所にコメントを追加

0b5vr commented on packages/three-vrm-core/src/expressions/VRMExpressionMaterialColorBind.ts:
```diff
@@ -21,9 +24,9 @@ export class VRMExpressionMaterialColorBind implements VRMExpressionBind {
       color: 'color',
       emissionColor: 'emissive',
-      outlineColor: 'outlineFactor',
-      rimColor: 'rimFactor',
-      shadeColor: 'shadeFactor',
+      outlineColor: 'outlineColorFactor',
+      rimColor: 'parametricRimColorFactor',
+      shadeColor: 'shadeColorFactor',
```
materialBindが対象とするMToonMaterialのプロパティがお名前がズレていたため、それの修正です

0b5vr replied:
今回、こんな感じで4つほどお名前ステークホルダーがいたのですが:

- VRM0.xのmaterial bindのお名前（undocumented）: https://github.com/pixiv/three-vrm/pull/1020#discussion_r945570408
- VRM1.0のmaterial bindのお名前: https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_vrm-1.0-beta/expressions.ja.md#materialcolorbind
- VRMC_materials_mtoonのプロパティ名: https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_materials_mtoon-1.0-beta/README.ja.md
- three-vrmのMToonMaterialのプロパティ名: https://github.com/pixiv/three-vrm/blob/1.0/packages/three-vrm-materials-mtoon/src/MToonMaterial.ts

ここのObjectは、このうちVRM1.0のmaterialBindの名前をthree-vrmのMToonMaterialのプロパティ名へマップするObjectです。

0b5vr commented on packages/three-vrm-materials-mtoon/src/shaders/mtoon.frag:
```diff
@@ -464,11 +465,12 @@ void main() {
   #ifdef USE_SHADEMULTIPLYTEXTURE
+    vec2 shadeMultiplyTextureUv = ( shadeMultiplyTextureUvTransform * vec3( uv, 1 ) ).xy;
     #if THREE_VRM_THREE_REVISION >= 137
-      material.shadeColor *= texture2D( shadeMultiplyTexture, uv ).rgb;
+      material.shadeColor *= texture2D( shadeMultiplyTexture, shadeMultiplyTextureUv ).rgb;
     #else
       // COMPAT: pre-r137
-      material.shadeColor *= shadeMultiplyTextureTexelToLinear( texture2D( shadeMultiplyTexture, uv ) ).rgb;
+      material.shadeColor *= shadeMultiplyTextureTexelToLinear( texture2D( shadeMultiplyTexture, shadeMultiplyTextureUv) ).rgb;
```
もともと `shadeMultiplyTextureUvTransform` が全く機能していなかったので（今回の趣旨と関連はするが直交するバグ）、それの修正です。

0b5vr reviewed (approved):
LGTM。ありがとうございました。
